### PR TITLE
Specification Design

### PR DESCRIPTION
**Pull request by lepture**
Review for GB2260 specification.

@cn/core @cn/programmers 

View in preview mode: https://github.com/cn/GB2260/blob/api-design/spec.md

### CODE REVIEW

**Comment by tonyseek** [comment]
LGTM :+1:

File: spec.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+A description of current division. e.g.
+
+```
+北京市 市辖区 崇文区
+```
+
+## Methods
+
+Methods the interface **GB2260** MUST contains.
```
Comment (lepture) [inline spec.md]: typo: contains

File: spec.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+// =>
+{
+  "code": "110103",
+  "name": "崇文区",
+  "year": null
+}
+```
+
+
+### `GB2260(year).provinces()` 省
```
Comment (tonyseek) [inline spec.md]: I prefer to stateless methods like `gb2260.Division.get_provinces(year=None)`. But `GB2260(year)` is acceptable too.

**Comment by lepture** [comment]
@tonyseek Updated with instance methods descriptions.

**Comment by Fleurer** [comment]
lgtm :+1:

**Comment by tonyseek** [comment]
LGTM for revision

**Comment by tonyseek** [comment]
@lepture We should add revision id, date and other meta information to the top of article.

File: spec.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+If `year` is not specified, use the latest data.
+
+
+### `.get(code)`
+
+Return the division of the given code.
+
+```javascript
+GB2260().get(110103)
```
Comment (gaomd) [inline spec.md]: The example should be `gb2260.get(110103)` or `(new GB2260(year)).get(110103)`.

**Comment by lepture** [comment]
@gaomd thanks. fixed.

File: spec.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+A description of current division. e.g.
+
+```
+北京市 市辖区 崇文区
+```
+
+## Methods
+
+Methods the interface **GB2260** MUST have. Create a new instance:
```
Comment (gaomd) [inline spec.md]: Requires better wording. 

e.g.: The interface **MUST** have a constructor method, to create a new instance:

File: spec.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+### `.provinces()` 省
+
+Return a list of provinces in `Division` data structure.
+
+
+### `.prefectures(province_code)` 地级市
+
+Return a list of prefecture level cities in `Division` data structure.
+
+A `province_code` is a 6-length province code. It can also be:
```
Comment (gaomd) [inline spec.md]: I think we should tell that the data type of `province_code` should be represented in **string** type.

Comment (lepture) [inline spec.md]: @gaomd it can be int.

Comment (gaomd) [inline spec.md]: @lepture I've update the question right before your reply.

---

Let's start over, we can only have one parameter type in strong typed language (like Java).

File: spec.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+## Interface
+
+Namespace should always be **GB2260**.
+
+## Data Structure
+
+There is a standard structure named `Division`. Those following properties
+should be included in it.
+
+- `code`: The six-digit number of the specific administrative division.
```
Comment (gaomd) [inline spec.md]: I suggest add data type to field descriptions, especially the `code` field, because `code` usually means `int`.

Comment (tonyseek) [inline spec.md]: > I suggest add data type to field descriptions

+1

File: spec.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
+# Specification
+
+Implement in your favorate languages by following this specification.
```
Comment (gaomd) [inline spec.md]: Typo: `favorate`.

File: spec.md (hunk 5 above)
Comment (tonyseek) [inline spec.md]: I suggest to let `code` always be text type in public interface. The `code` is a identity field, not a value could be calculated.

**Comment by lepture** [comment]
@gaomd updated with data type

File: spec.md (hunk 5 above)
Comment (gaomd) [inline spec.md]: +1 `string` type as the declared parameter type.

Comment (lepture) [inline spec.md]: It should be str now.

File: spec.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+Namespace should always be **GB2260**.
+
+## Data Structure
+
+There is a standard structure named `Division`. Those following properties
+should be included in it.
+
+- `code`: (str) The six-digit number of the specific administrative division.
+- `name`: (str) The Chinese name of the specific administrative division.
+- `year`: (str) Optional. The revision year, and empty means "latest".
```
Comment (gaomd) [inline spec.md]: > 1. Those following properties **should** be included in it.
> 2. - `year`: (str) **Optional.** The revision year, and empty means "latest".

I think the bolded words have conflicts, and not describing `null` situation.

I suggest change **should** to **must**, then remove **Optional.**, and change **empty** to `null`.

That is:

> 1. Those following properties must be included in it.
> 2. - `year`: (str|null) The revision year, and `null` means "latest".

File: spec.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+```javascript
+{
+  "code": "110103",
+  "name": "崇文区",
+  "year": "2003"
+}
+```
+
+The administrative level representation depends on recursion records. The
+division codes `XX0000` (`XX` is non-zero digits) mean provinces. And there are
```
Comment (gaomd) [inline spec.md]: According to the context, `XX` is a plural form, the correct wording may be:

> `XX` **are** non-zero digit**s**

Actually, this makes a lot more sense to me:

> `X` is non-zero digit

**Pull request closed without merging**